pretyping/evarsolve.ml

```diff
@@ -1253,7 +1253,9 @@ let rec is_constrainable_in top env evd k (ev,(fv_rels,fv_ids) as g) t =
       Array.for_all (is_constrainable_in false env evd k g) params
   | Ind _ -> Array.for_all (is_constrainable_in false env evd k g) args
   | Prod (na,t1,t2) -> is_constrainable_in false env evd k g t1 && is_constrainable_in false env evd k g t2
-  | Evar (ev',_) -> top || not (Evar.equal ev' ev) (*If ev' needed, one may also try to restrict it*)
+  | Evar (evk',_ as ev') ->
+    (*If ev' needed, one may also try to restrict it*)
+    top || not (Evar.equal evk' evk || occur_evar evd evk (Evd.existential_type evd ev'))
```

Comment:
This is not a simple cleanup is it?

Reply:
Yes, this is one of the commits with a semantic change. I'll put those in another PR.